### PR TITLE
compiler: add alloc attributes to runtime.alloc

### PR DESCRIPTION
**Pull request by aykevl**
This gives a small improvement now, and is needed to be able to use the Heap2Stack transform that's available in the Attributor pass. This Heap2Stack transform could replace our custom OptimizeAllocs pass.

Most of the changes are just IR that changed, the actual change is relatively small.

To give an example of why this is useful, here is the code size before this change:

    $ tinygo build -o test -size=short ./testdata/stdlib.go
       code    data     bss |   flash     ram
      95620    1812     968 |   97432    2780

    $ tinygo build -o test -size=short ./testdata/stdlib.go
       code    data     bss |   flash     ram
      95380    1812     968 |   97192    2780

That's a 0.25% reduction. Not a whole lot, but nice for such a small patch.

### CODE REVIEW

**Comment by aykevl** [comment]
Looks like this is failing under LLVM 14.

**Comment by aykevl** [comment]
LLVM 14 regression should be fixed now.

**Comment by deadprogram** [comment]
Interestingly, this PR causes the `itsybitsy-nrf52840` to fail the TinyHCI. I verified the HW, and then noticed that other PRs were able to still pass.

See https://github.com/tinygo-org/tinygo/pull/3540/checks?check_run_id=11934025697

**Comment by aykevl** [comment]
I ran the entire test corpus, which passed. So if it's a real bug, it's a very specific one.

I also built the itsybitsy-nrf52840 TinyHCI binary locally and there is a difference in the binary output, so a real bug triggered by this change is plausible.

**Comment by aykevl** [comment]
I can reproduce the issue locally.

**Comment by aykevl** [comment]
Huh, there is indeed a rather important bug: somehow the optimizer decided that it's fine to just return 0 instead of the actual value.

**Comment by aykevl** [comment]
Yup, reducing further the compiler converts this line:

```llvm
  %11 = load i16, ptr %value, align 2
  %12 = shl i16 %11, 4
  ret i16 %12
```

Into this:

```llvm
  ret i16 0
```

(`%value` here is `var value int16` from the beginning of the `.Get()` function, and these lines are the IR of the last lines of `.Get()`: `return uint16(value << 4)`).

The reasoning appears to be that `runtime.alloc` returns a zeroed allocation (correct), and there is no store to this before the load at the end (not entirely correct). The problem is that while there is no store, the pointer is stored into a register which captures the pointer (and can therefore lead to arbitrary changes to the value).
In other words, this appears like a LLVM bug.

**Comment by aykevl** [comment]
Reported the bug upstream: https://github.com/llvm/llvm-project/issues/61362

I'll convert this to a draft until the bug is resolved in some way.

**Comment by aykevl** [comment]
Fixed the issue for the nrf. I looked for other issues like this, and there may be some, but not as clear as this one.
(To be clear, the code has always been broken, the optimizer just didn't catch it until we added the alloc attributes).

**Comment by deadprogram** [comment]
Have not been able to get this PR to pass the CircleCI with the older version compatibility tests. Perhaps it in fact is not compatible?

**Comment by deadprogram** [comment]
Perhaps rebase against `dev` might convince it to pass?

**Comment by aykevl** [comment]
No, I can reproduce it locally. This looks more like a bug in the PR.

**Comment by aykevl** [comment]
Whoops, that's what you get from working with two different git repos. I force-pushed an older branch which re-introduced the LLVM 14 bug.
With the latest push, it should finally be fixed.

**Comment by deadprogram** [comment]
All green at last, thank you very much @aykevl now merging.
